spec/unit/hanami/validations/predicates/schema/custom_spec.rb

```diff
@@ -120,6 +120,108 @@ def self.name
     end
   end
 
+  describe 'with custom predicate within predicates block followed by a predicates module' do
+    before do
+      @validator = Class.new do
+        include Hanami::Validations
+
+        def self.name
+          'Validator'
+        end
+
+        predicate :url?, message: 'must be an URL' do |current|
+          current.start_with?('http')
+        end
+
+        predicates(
+          Module.new do
+            include Hanami::Validations::Predicates
+            self.messages_path = "spec/support/fixtures/messages.yml"
+
+            predicate(:email?) do |current|
+              current.match(/@/)
+            end
+          end
+        )
+
+        validations do
+          required(:foo) { url? | email? }
+        end
+      end
+    end
+
+    describe 'with valid ulr input' do
+      let(:input) { { foo: 'http://hanamirb.org' } }
+
+      it 'is successful' do
+        expect_successful result
+      end
+    end
+
+    describe 'with valid email input' do
+      let(:input) { { foo: 'foo@mailinator.com' } }
+
+      it 'is successful' do
+        expect_successful result
+      end
+    end
+
+    describe 'with invalid input' do
+      let(:input) { { foo: 'test' } }
+
+      it 'is not successful' do
+        expect_not_successful result, ['must be an URL or must be an email']
+      end
+    end
+  end
+
+  describe 'with custom predicate as predicates module followed by a predicates block' do
+    before do
+      @validator = Class.new do
+        include Hanami::Validations
+
+        def self.name
+          'Validator'
+        end
+
+        predicates(
+          Module.new do
+            include Hanami::Validations::Predicates
+            self.messages_path = "spec/support/fixtures/messages.yml"
+
+            predicate(:email?) do |current|
+              current.match(/@/)
+            end
+          end
+        )
+
+        predicate :url?, message: 'must be an URL' do |current|
+          current.start_with?('http')
+        end
+
+        validations do
+          required(:foo) { url? | email? }
+        end
+      end
+    end
+
+    describe 'with valid input' do
+      let(:input) { { foo: %w[http://hanamirb.org foo@mailinator.com].sample } }
```

Comment:
One more place with `sample` in tests here.

Reply:
Address here 7530bc38092d0d1cd631c2583d6285aceee88c61 and apologies